### ISSUE TITLE
Discrepancy between code & manual calculation of parameter gradient w.r.t older version of parameter

### ISSUE THREAD

**Issue by hazrmard** [facebookresearch/higher#95]
**Summary**: Using `torch.autograd.grad` to calculate gradient of parameter at time 1 w.r.t time 0 yields different results against manual computation.

---
**Manual**

Take this function where `a_[0],b_[0]` are current parameters (t=0), `x` is input and `y_` is  predicted model output. `a, b` and `y` are true parameters and value respectively.

    y_ = (a_[0]**2 + b_[0]**2) * x

The loss is 

    L = 0.5 * (y_ - y) ** 2

Here, I consider just `a`. The gradients are:

    d L  / d y_  = y_ - y                   # loss w.r.t prediction
    d y_ / d a_[0] = 2 a_[0] x              # prediction w.r.t parameter
    d L  / d a_[0] = (y_ - y)(2 a_[0] x)    # loss w.r.t parameter via chain rule

I am using SGD with learning rate=1 for simplicity. So, the next version of parameter `a`, `a_[1]` will be:

    a_[1] = a_[0] -  d L  / d a_[0]        # update opposite direction of gradient
    a_[1] = a_[0] - (y_ - y)(2 a_[0] x)    # substituting
    a_[1] = a_[0] * (1 - (y_ - y)(2 x) )   # factoring out a_[0]
    a_[1] = a_[0] * (1- 2 x y_ + 2 x y)

Which gives the gradient between different versions of `a_` as:

    d a_[1] / d a_[0] = 1- 2 x y_ + 2 x y

Assuming true `a=3, b=4`, and current parameters `a_[0]=2, b_[0]=3`, and input `x=0.1`, then

    y =  (3**2 + 4**2) * 0.1 = 2.5
    y_ = (2**2 + 3**2) * 0.1 = 1.3
    L = 0.5 (1.3 - 2.5)**2 = 0.72

    d L  / d y_ = -1.2
    d y_ / d a_[0] = 2 * 2 * 0.1 = 0.4
    d L  / d a_[0] = (1.3 - 2.5)(2 * 2 * 0.1) = -1.2 * 0.4 = -0.48

    a_[1] = 2 - (-0.48) = 2.48
    d a_[1] / d a_[0] = 1 - 2 * 0.1 * 1.3 + 2 * 0.1 * 2.5 = 1 - 0.26 + 0.5 = 1.24

---
**Code**

```python
# torch==1.5.1
# higher==0.2.1
from torch import nn, optim
import higher
from higher import innerloop_ctx
from higher.optim import DifferentiableSGD

class Model(nn.Module):
    """Implements the function y = (a**2 + b**2) * x"""
    
    def __init__(self, a=2., b=3.):
        """a,b are starting parameter values"""
        super().__init__()
        # Use nn.Modules
        self.a = nn.Linear(1,1,False)
        self.a.weight.data.fill_(a)
        self.b = nn.Linear(1,1,False)
        self.b.weight.data.fill_(b)
        # Or custom parameters (toggle comments in forward()):
        # self.a = nn.Parameter(torch.ones(1, requires_grad=True) * a)
        # self.b = nn.Parameter(torch.ones(1, requires_grad=True) * b)
    
    def forward(self, x):
        return self.a(self.a(x)) + self.b(self.b(x))    # with modules
        # return (self.a ** 2 + self.b ** 2) * x        # with custom


class Opt(DifferentiableSGD):
    """Using higher's SGD implementation, except that it prints out loss gradients w.r.t parameters"""
    
    def _update(self, grouped_grads, **kwargs) -> None:
        print('d Loss / d parameter(t=0): {:.3f} \t {:.3f}'.format(*[g.item() for g in grouped_grads[0]]))
        return super()._update(grouped_grads, **kwargs)
        
higher.register_optim(optim.SGD, Opt)


x = torch.ones(1) * 0.1
a, b = 3, 4    # true parameter values
m = Model()
o = optim.SGD(m.parameters(), lr=1.)
# Single training step t=0 ==> t=1
with higher.innerloop_ctx(m, o) as (hm, ho):
    y_ = hm(x)
    y = (a**2 + b**2) * x
    loss = 0.5 * (y_ - y) ** 2
    print('Loss:\t\t{:.3f}'.format(loss.item()))
    print('d Loss / d y_: {:.3f}'.format(torch.autograd.grad(loss, y_, retain_graph=True)[0].item()))
    print('Current params: \ta_[0]:{:.3f} \t b_[0]:{:.3f}'.format(
          *[p.item() for p in hm.parameters(time=0)]))
    
    ho.step(loss)
    
    print('Updated params: \ta_[1]:{:.3f} \t b_[1]:{:.3f}'.format(
          *[p.item() for p in hm.parameters(time=1)]))
    
    print('d a_[1] / d param(t=0): a_[0]:{:.3f} \t b_[0]:{:.3f}'.format(
          *[g.item() for g in \
            torch.autograd.grad(list(hm.parameters(time=1))[0],  # this is first param at t=1 (a_[1])
                                hm.parameters(time=0))]))        # w.r.t all params at t=0
```

This outputs:

```
Loss:		0.720
d Loss / d y_: -1.200
Current params: 	a_[0]:2.000 	 b_[0]:3.000
d Loss / d params(t=0): a_[0]:-0.480 	 b_[0]-0.720
Updated params: 	a_[1]:2.480 	 b_[1]:3.720
d a_[1] / d param(t=0): a_[0]:1.080 	 b_[0]:-0.240
```

As you can see, loss gradients and outputs match between manual and code versions. But the `autograd` calculation of `d a_[1] / d a_[0]` vs manual is 1.08 vs 1.24. Is there a reason for this discrepancy?

**Comment by egrefen** [comment]
@hazrmard Can you explain why you closed the issue? Was it working as intended?

**Comment by hazrmard** [comment]
> 
> 
> @hazrmard Can you explain why you closed the issue? Was it working as intended?

It was an error in my math at the exact spot where I noted the discrepancy. While taking the derivative w.r.t `a_[0]` of

    a_[1] = a_[0] * (1- 2 x y_ + 2 x y)

I forgot that `y_` is a function of `a_[0]` and didn't expand it during the derivation.
